Content.Server/Automod/AutomodManager.cs

```diff
@@ -0,0 +1,260 @@
+﻿using System.Diagnostics;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Content.Server.Database;
+using Content.Shared.Automod;
+using Content.Shared.Database;
+using Robust.Shared.Player;
+using Robust.Shared.Prototypes;
+
+namespace Content.Server.Automod;
+
+public sealed class AutomodManager : IAutomodManager, IPostInjectInit
+{
+    [Dependency] private readonly IPrototypeManager _protoMan = default!;
+    [Dependency] private readonly IEntityManager _entMan = default!;
+    [Dependency] private readonly ILogManager _logMan = default!;
+    [Dependency] private readonly IServerDbManager _db = default!;
+
+    private ISawmill _log = default!;
+
+    // Filters
+    private readonly Dictionary<AutomodTarget, Dictionary<Regex, AutomodFilterDef>> _regexFilters = new();
+
+    #region Initialize
+
+    public void Initialize()
+    {
+        IoCManager.InjectDependencies(this);
+
+        LoadAutomodFiltersFromDatabase();
+    }
+
+    private async void LoadAutomodFiltersFromDatabase()
+    {
+        var filters = await _db.GetAllAutomodFiltersAsync();
+        foreach (var filter in filters)
+        {
+            AddFilter(filter);
+        }
+    }
+
+    void IPostInjectInit.PostInject()
+    {
+        _log = _logMan.GetSawmill("automod");
+    }
+
+    #endregion
+
+    #region Filter
+
+    public bool Filter(AutomodTarget target, string inputText, ICommonSession session)
+    {
+        return RegexFilter(target, inputText, session);
+    }
+
+    private bool RegexFilter(AutomodTarget target, string inputText, ICommonSession session)
+    {
+        // Ensure that only 1 bit is set
+        Debug.Assert(target != 0 && (target & (target - 1)) == 0);
+
+        var blocked = true;
+
+        // No filters defined for target
+        if (!_regexFilters.TryGetValue(target, out var automodFilters))
+            return true;
+
+        foreach (var (regex, filter) in automodFilters)
+        {
+            var regexMatches = regex.Matches(inputText);
+            if (regexMatches.Count == 0)
+                continue;
+
+            var textMatches = new Dictionary<string, int>();
```

Comment:
This should be a flat list.

Reply:
Made it a tuple list. [9b54255](https://github.com/space-wizards/space-station-14/pull/28570/commits/9b54255a94be6934e1b5c1a876a68ee928c07543)